controllers/storagecluster/odfinfoconfig.go

```diff
@@ -0,0 +1,294 @@
+	odfDeploymentTypeExternal = "external"
+	odfDeploymentTypeInternal = "internal"
+	rookCephMonSecretName     = "rook-ceph-mon"
+	fsidKey                   = "fsid"
+	odfOperatorNamePrefix     = "odf-operator"
+	odfInfoConfigMapName      = "odf-info"
+	odfInfoMapKind            = "ConfigMap"
+)
+
+type odfInfoConfig struct{}
+
+var mutex sync.RWMutex
+
+// ensureCreated ensures that a ConfigMap resource exists with its Spec in
+// the desired state.
+func (obj *odfInfoConfig) ensureCreated(r *StorageClusterReconciler, storageCluster *ocsv1.StorageCluster) (reconcile.Result, error) {
+	operatorNamespace, err := util.GetOperatorNamespace()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	odfInfoConfigMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      odfInfoConfigMapName,
+			Namespace: operatorNamespace,
+		},
+	}
+
+	mutex.Lock()
+	defer mutex.Unlock()
+	_, err = ctrl.CreateOrUpdate(r.ctx, r.Client, odfInfoConfigMap, func() error {
+		if err = ctrl.SetControllerReference(storageCluster, odfInfoConfigMap, r.Scheme); err != nil {
+			return err
+		}
+		r.Log.Info("Creating or updating odf-info config map", odfInfoMapKind, klog.KRef(storageCluster.Namespace, odfInfoConfigMapName))
+		odfInfoKey := obj.getOdfInfoKeyName(storageCluster)
+		var data map[string]string
+		if err = yaml.Unmarshal([]byte(odfInfoConfigMap.Data[odfInfoKey]), data); err != nil {
+			return fmt.Errorf("failed to get odf-info config map data: %v", err)
+		}
+
+		odfInfoData, configErr := getOdfInfoData(r, storageCluster)
+		if configErr != nil {
+			return fmt.Errorf("failed to get odf-info config map data: %v", configErr)
+		}
+		if odfInfoConfigMap.Data == nil {
+			odfInfoConfigMap.Data = map[string]string{}
+		}
+		// Creates or appends to the data map
+		odfInfoConfigMap.Data[odfInfoKey] = odfInfoData
+		return nil
+	})
+	if err != nil {
+		r.Log.Error(err, "failed to create or update odf-info config map", odfInfoMapKind, klog.KRef(storageCluster.Namespace, odfInfoConfigMapName))
+		return reconcile.Result{}, fmt.Errorf("failed to create or update odf-info config: %v", err)
+	}
+	return reconcile.Result{}, nil
+}
+
+func (obj *odfInfoConfig) getOdfInfoKeyName(storageCluster *ocsv1.StorageCluster) string {
+	return fmt.Sprintf("%s/%s.%s", storageCluster.Namespace, storageCluster.Name, odfInfoKeySuffix)
+}
+
+// ensureDeleted is dummy func for the odfInfoConfig
+func (obj *odfInfoConfig) ensureDeleted(r *StorageClusterReconciler, storageCluster *ocsv1.StorageCluster) (reconcile.Result, error) {
+	operatorNamespace, err := util.GetOperatorNamespace()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	odfInfoConfigMap := &corev1.ConfigMap{}
+	odfInfoConfigMap.Name = odfInfoConfigMapName
+	odfInfoConfigMap.Namespace = operatorNamespace
+	if err = r.Client.Get(r.ctx, client.ObjectKeyFromObject(odfInfoConfigMap), odfInfoConfigMap); err != nil && !errors.IsNotFound(err) {
+		return reconcile.Result{}, err
+	}
```

Comment:
The two cases need to be separated:
1. we were not able to load the CM because it does not exist - in this case, we want to return from the function with a successful result because the intent of the function is to remove the key from the CM and if there is no CM then there is nothing to do.
2. There was a different error - return with an error


```suggestion
	if err = r.Client.Get(r.ctx, client.ObjectKeyFromObject(odfInfoConfigMap), odfInfoConfigMap); err != nil {
	        if errors.IsNotFound(err) {
			return reconcile.Result{}, nil
		}
		return reconcile.Result{}, err
	}
```